README.md
#Tumor-resilient T-cell model  

**Prerequisite**:  
1, [ridge_significance](https://github.com/data2intelligence/ridge_significance)    
Please read its README.md and run test to make sure the successful installation.    
  
2, [CytoSig](https://github.com/data2intelligence/CytoSig)  
Please read its README.md and run test to make sure the successful installation.  
  
3, pandas >= 1.1.4: You may install [anaconda](https://www.anaconda.com) to include all required python packages.  


**Install**:
python setup.py install  


**Test**:  
1, Run the Tres model to compute signatures of tumor-resilient T cells  
python -m unittest tests.run  

2, Run the T-cell efficacy prediction using a pre-computed Tres signature  
python -m unittest tests.predict  


**Usage 1: compute Tres gene scores**:      
Tres.py -i <single-cell data> -o output -c <count threshold, default: 100> -n <mean-normalization, default: 0>
  
1, input: single-cell transcriptomic data matrix. The format could be a text file, gzip file, or pickle file. Please see "Melanoma.GSE115978.post\_Ipi.CD8.gz" as an example.  
  
Column: Cell ID in the format of Cell\_Type.Patient\_ID.Cell\_Barcode (or any sample ID).  
Row: Human gene symbols.  
Value: Gene expression value. We suggest log2(TPM/10 + 1) normalized by subtracting the mean from all conditions.      
  
Note: For 10x genomics data, we also included a command line tool **convert_mtx.py** to convert cell ranger output format to Tres input format. Please read the usage section at the end.  

2, output: results from the interaction test.  
  
Column: Condition ID in the format of Value\_Type.Cell\_Type.Patient\_ID.Signal\_ID  
  -Value\_Type: t: t-value from the student t-test on the interaction term in regression, p: p-value of the associated t-value  
  -Cell\_Type: the same as input  
  -Patient\_ID: the same as input  
  -Signal\_ID: TGFB1, TRAIL, PGE2, representing immunosuppressive signals in the tumor microenvironment  
  
Row: Gene ID  
  
3, count threshold: the minimum number of cells in a sample to be analyzed. For example, with the default value 100, Tres will ignore all combinations of cell type and patient with less than 100 single cells sequenced.  

4, mean-normalization: whether normalize input gene expression values by subtracting the mean value across all conditions. We assume the users should do similar normalization before using Tres, thus the default value is 0. Otherwise, please activate this option.   
  
Example:  
In the directory of README.md, please type: Tres.py -i tests/Melanoma.GSE115978.post\_Ipi.CD8.gz -o output  


**Usage 2: compute sample-wise correlations with pre-computed Tres signature**:      
Tres.py -m predict -i <sample data> -o output -c <count threshold, default: 100> -n <mean-normalization, default: 0>  
  
1, input: sample transcriptomic data matrix. The format could be a text file, gzip file, or pickle file. Please see "Sample\_Melanoma\_GSE120575.gz" as an example.  
  
Column: Sample ID.  
Row: Human gene symbols.  
Value: Gene expression value. We suggest log2(TPM + 1) normalized by subtracting the mean from all conditions.      
  
2, output: correlation and p-values for each sample.  
  
Column:   
  r: Pearson correlation  
  p: p-value from two-sided student t-test    
  
Row: Sample ID  
  
3, count threshold: the minimum number of genes in overlap between input gene expression matrix and the pre-computed Tres signature.  

4, mean-normalization: whether normalize input gene expression values by subtracting the mean value across all conditions. We assume the users should do similar normalization before using Tres, thus the default value is 0. Otherwise, please activate this option.  
  
Example:  
In the directory of README.md, please type: Tres.py -m predict -i tests/Sample\_Melanoma\_GSE120575\_Tex.gz -o output  

---  
  
#Convert cellranger output to Tres input format:  
The input to Tres is a dense matrix file. However, the 10x cellranger output is a feature-barcode matrix in sparse format. Thus, we provide a command line tool for format conversation.  
The converter will merge all files with prefix provided after removing cell barcodes and genes with low read counts, transforming read count to log2(TPM/10+1), and subtracting the average value for each gene across all cell barcodes.  

**Prerequisite**:    
scipy >= 1.7.1, pandas >= 1.1.4: You may install [anaconda](https://www.anaconda.com) to include all required python packages.  

**Install**:  
This command line tool will be installed together with Tres.  
  
**Test**:  
First, download some sample cellranger output files:  
wget [https://hpc.nih.gov/~Jiang_Lab/Tres/GSE139829_sample.tar.gz](https://hpc.nih.gov/~Jiang_Lab/Tres/GSE139829_sample.tar.gz)  
tar xvf GSE139829_sample.tar.gz  
  
Second, cd into the extracted folder and type:  
convert\_mtx.py -i GSM4147093\_UMM059\_,GSM4147096\_UMM063\_,GSM4147099\_UMM066\_ -o GSE139829_density -s Cytotoxic\ CD8

Then, you can see a converted file "GSE139829_density.pickle.gz", on which you can apply Tres by:  
Tres.py -i GSE139829\_density.pickle.gz -o GSE139829\_sample\_Tres  

**Usage**:  
convert_mtx.py -i <input list, separated by ,> -o <output prefix> -c <barcode count threshold, default: 1000> -f <gene non-zero fraction threshold, default: 0.95> -s <cell subsets, default: leave empty>  

1, input list:  
Please input a list of file prefix of cell ranger output separated by ",". For example, f1,f2, .... For each prefix, we assume barcode, gene, and sparse matrix files as: fprefix + 'barcodes.tsv.gz', fprefix + 'genes.tsv.gz', fprefix + 'matrix.mtx.gz'.  
To generate Tres input file, we also need a cell annotation file named as fprefix + 'cell_map.gz'.  
The first column should contain cell barcodes in the barcode file.  
The second column should has a name "Sample", representing the patient sample ID.  
The third column should has a name "Cell Type", representing the cell type annotation.  
  
2, output: Converted file in python pickle format.  

3, barcode count threshold:  
The minimum number of read count for a cell barcode. If a cell barcode contains very few read counts below the threshold, the converter will drop the cell barcode.  

4, gene non-zero fraction:  
The maximum fraction of zero read count (dropout) for a gene. If a gene gets zero read count (dropout) in most cell barcodes, the converter will drop the gene.  

5, cell subset:  
The T-cell labels to be analyzed in Tres. If nothing is provided, Tres will analyze all cell types.   


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.  

- 2021-11-01: formal software release  
- 2021-12-08: add a commandline tool convert_mtx.py to convert cellranger output files to Tres input format.  


test.sh
#!/bin/sh
python Tres.py -i Melanoma.GSE115978.post_Ipi.CD8.pickle.gz -o output


tests/run.py
import unittest, os, pandas, pathlib, subprocess

fpath = pathlib.Path(__file__).parent.absolute()

output = os.path.join(fpath, 'output')

class TestRun(unittest.TestCase):
    def max_difference(self, A, B, ratio_thres):
        self.assertTrue(A.shape == B.shape)
        
        # weak similarity test because different machines really give different results.
        self.assertTrue((A-B).abs().median().max() < 1-ratio_thres)

    def test_run(self):
        data = os.path.join(fpath, 'Melanoma.GSE115978.post_Ipi.CD8.gz')
        result_expected = pandas.read_csv(output + '.run.gz', sep='\t', index_col=0)
        
        # run command line as usage 1
        cmd = ['Tres.py', '-i', data, '-o', output]
        
        p = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
        p.communicate()
        
        result = pandas.read_csv(output, sep='\t', index_col=0)
        os.remove(output)
        
        # a known issue: depending on the computer, passed gene will be slightly different
        common = result.index.intersection(result_expected.index)
        
        ratio_thres = 0.95
        self.assertTrue(common.shape[0]/result.shape[0] > ratio_thres)
        self.assertTrue(common.shape[0]/result_expected.shape[0] > ratio_thres)
        
        # make sure the two usages get the same result   
        self.max_difference(result.loc[common], result_expected.loc[common], ratio_thres)

if __name__ == '__main__': unittest.main()


tests/predict.py
import unittest, os, pandas, pathlib, subprocess

fpath = pathlib.Path(__file__).parent.absolute()

output = os.path.join(fpath, 'output')

class TestPredict(unittest.TestCase):
    def max_difference(self, A, B):
        self.assertTrue(A.shape == B.shape)
        self.assertTrue((A-B).abs().max().max() < 1e-8)

    def test_predict(self):
        data = os.path.join(fpath, 'Sample_Melanoma_GSE120575.gz')
        result_expected = pandas.read_csv(output + '.predict.gz', sep='\t', index_col=0)
        
        # run command line as usage 1
        cmd = ['Tres.py', '-m', 'predict', '-i', data, '-o', output]
        
        p = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
        p.communicate()
        
        result = pandas.read_csv(output, sep='\t', index_col=0)
        os.remove(output)
        
        # make sure the two usages get the same result   
        self.max_difference(result, result_expected)

if __name__ == '__main__': unittest.main()


src/Tres.py
#!/usr/bin/env python

import time
import resource
import os, sys, pandas, numpy, pathlib, getopt
import CytoSig

from scipy import stats
from statsmodels.stats.multitest import multipletests

err_tol = 1e-8
verbose_flag = False

fpath = pathlib.Path(__file__).parent.absolute()

pivots = ['TGFB1', 'TRAIL', 'PGE2']

def profile_geneset_signature(expression):
    signature = []
    fin = open(os.path.join(fpath, 'Tres.kegg'))
    for l in fin:
        fields = l.strip().split('\t')
        
        s = fields[2:]
        signature.append(pandas.Series(numpy.ones(len(s)), index=s, name=fields[0]))
    fin.close()
    
    signature = pandas.concat(signature, axis=1, join='outer', sort=False)
    signature.fillna(0, inplace=True)
    
    common = expression.index.intersection(signature.index)
    signature, expression = signature.loc[common], expression.loc[common]
    
    background = signature.mean(axis=1)
    background.name = 'study bias'
    
    X = signature.loc[:, ['KEGG_CELL_CYCLE', 'KEGG_DNA_REPLICATION']].mean(axis=1)
    X.name = 'Proliferation'
    
    X = pandas.concat([background, X], axis=1, join='inner')
    
    result = CytoSig.ridge_significance_test(X, expression, alpha=0, alternative="two-sided", nrand=0, verbose=verbose_flag)
    
    return result[2].loc['Proliferation']




def interaction_test(expression, X, y):
    signal = X.loc[:, 'pivot']
    
    failed = []
    merge = []
    
    for gid, arr in expression.iterrows():
        #if gid not in ['FIBP', 'TMEM222', 'IL7R']: continue
        
        X.loc[:,'partner'] = arr
        X.loc[:,'interaction'] = arr * signal
    
        # other covariates have no sufficient variation
        if arr.std() < err_tol or X.loc[:,'interaction'].std() < err_tol: continue
        
        try:
            y = pandas.DataFrame(y)
            result = CytoSig.ridge_significance_test(X, y, alpha=0, alternative="two-sided", nrand=0, flag_normalize=False, verbose=verbose_flag)
        
        except ArithmeticError:
            failed.append(gid)
            continue
        
        tvalue = result[2].loc['interaction'].iloc[0]
        pvalue = result[3].loc['interaction'].iloc[0]
        
        merge.append(pandas.Series([tvalue, pvalue], index=['t', 'p'], name=gid))    
    
    result = pandas.concat(merge, axis=1, join='inner').transpose()
    result['q'] = multipletests(result['p'], method='fdr_bh')[1]
    
    return result


def prediction(expression, output_file, count_thres):
    signature = pandas.read_csv(os.path.join(fpath, 'Tres.prediction_signature.gz'), sep='\t', index_col=0)['Tres']
    
    common = expression.index.intersection(signature.index)
    if common.shape[0] < count_thres:
        sys.stderr.write('Low overlap between input data genes and Tres signature genes.\n')
        sys.exit(1)
    
    print(common.shape[0], 'overlapping genes')
    
    expression = expression.loc[common]
    signature = signature.loc[common]
    
    merge = []
    
    try:
        for title, arr in expression.items():
            r, p = stats.pearsonr(arr, signature)
            merge.append(pandas.Series([r, p], index=['r', 'p'], name=title))
    except:
        sys.stderr.write('Pearson correlation failed.\n')
        sys.exit(1)
    
    result = pandas.concat(merge, axis=1, join='outer').transpose().sort_values('r')
    result.to_csv(output_file, sep='\t', index_label='ID')




def run_Tres(expression, output_file, count_thres):
    """
    run Tres model to get the signature
    """
    
    ###############################################################
    # read signature
    signature = os.path.join(sys.prefix, 'bin', 'signature.centroid.expand')
    
    if not os.path.exists(signature):
        sys.stderr.write('Cannot find signature file %s. Please make sure your CytoSig installation is successful.\n' % signature)
        sys.exit(1)
    else:
        signature = pandas.read_csv(signature, sep='\t', index_col=0)
    
    ###############################################################
    # compute signaling activity
    try:
        result_signaling = CytoSig.ridge_significance_test(signature, expression, alpha=1E4, verbose=verbose_flag)
    
    except ArithmeticError:
        sys.stderr.write('CytoSig regression failed.\n')
        sys.exit(1)
    
    # get the z-scores
    result_signaling = result_signaling[2]
    
    ###############################################################
    # compute proliferation signature
    try:
        result_prolifertion = profile_geneset_signature(expression)
    
    except ArithmeticError:
        sys.stderr.write('KEGG proliferation computation failed.\n')
        sys.exit(1)
    
    ###############################################################
    # Interaction test
    flag_group = ['.'.join(v.split('.')[:2]) for v in expression.columns]
    
    expression_group = expression.groupby(flag_group, axis=1)
    
    merge = []
    
    for title, expression_sub in expression_group:
        N = expression_sub.shape[1]
        if N < count_thres: continue
        
        print('process', title, N)
        
        # remove rows all zeros
        flag_nonzero = (expression_sub == 0).mean(axis=1) < 1
        if sum(~flag_nonzero) > 0: expression_sub = expression_sub.loc[flag_nonzero]
        
        y = result_prolifertion.loc[expression_sub.columns]
        
        # regression scaffold
        X = pandas.DataFrame(numpy.zeros((N,4)),
            columns = ['const', 'pivot', 'partner', 'interaction'],
            index = expression_sub.columns)
        
        X.loc[:, 'const'] = 1
        
        for pivot in pivots:
            X.loc[:,'pivot'] = result_signaling.loc[pivot, expression_sub.columns]
        
            result = interaction_test(expression_sub, X, y)
            result.columns += ('.%s.%s' % (title, pivot))
            merge.append(result)
    
    result = pandas.concat(merge, axis=1, join='outer')
    result.to_csv(output_file, sep='\t', index_label='ID')



def main():
    run_mode = 'run'
    input_file = output_file = None
    count_thres = 100
    normalization = False
        
    prompt_msg = 'Usage:\nTres.py -m <running mode (run|predict), default: %s> -i <input single-cell data> -o <output prefix> -c <count threshold, default: %d> -n <mean-normalization, default: %d>\n' % (run_mode, count_thres, normalization)
    
    try:
        opts, _ = getopt.getopt(sys.argv[1:], "hm:i:o:c:n:", [])
    
    except getopt.GetoptError:
        sys.stderr.write('Error input\n' + prompt_msg)
        sys.exit(2)
    
    if len(opts) == 0:
        sys.stderr.write('Please input some parameters or try: Tres.py -h\n')
        sys.exit(2)
    
    for opt, arg in opts:
        if opt == '-h':
            print(prompt_msg)
            sys.exit()
        
        elif opt in ("-m"):
            run_mode = arg
            
            if run_mode not in ['run', 'predict']:
                sys.stderr.write('Cannot recognize run mode %s. Please only use run or predict.\n' % run_mode)
                sys.exit(1)
        
        elif opt in ("-i"):
            input_file = arg
        
        elif opt in ("-o"):
            output_file = arg
        
        elif opt in ("-n"):
            normalization = (int(arg) != 0)
        
        elif opt in ("-c"):
            try:
                count_thres = int(arg)
            except:
                sys.stderr.write('count threshold %s is not a valid integer number.\n' % arg)
                sys.exit(1)
            
            if count_thres < 10:
                sys.stderr.write('count threshold %d < 10. Please input a number >= 10\n' % count_thres)
                sys.exit(1)
    
    if input_file is None:
        sys.stderr.write('Please provide a input file\n')
        sys.exit(1)
    
    elif not os.path.exists(input_file):
        sys.stderr.write('Cannot find input file %s\n' % input)
        sys.exit(1)
    
    if output_file is None:
        output_file = input_file + '.Tres_output'
        sys.stderr.write('No output file input. Automatically generate one as %s\n' % output_file)
    
    ###############################################################
    # read input
    try:
        f = os.path.basename(input_file)
        
        if f.find('.pickle') >= 0:
            expression = pandas.read_pickle(input_file)
        else:
            expression = pandas.read_csv(input_file, sep='\t', index_col=0)
    except:
        sys.stderr.write('Fail to open input file %s\n' % input_file)
        sys.exit(1)
    
    # gene and sample names must be unique
    assert expression.index.value_counts().max() == 1
    assert expression.columns.value_counts().max() == 1
    
    print('input matrix dimension', expression.shape)
    
    if normalization:
        background_expression = expression.mean(axis=1)
        expression = expression.subtract(background_expression, axis=0)
    
    if run_mode == 'run':
        run_Tres(expression, output_file, count_thres)    
    
    elif run_mode == 'predict':
        prediction(expression, output_file, count_thres)
    
    else: # impossible to get here
        assert False
    
    return 0



if __name__ == '__main__':
    time_start = time.perf_counter()
    main()
    time_elapsed = (time.perf_counter() - time_start)
    memMb=resource.getrusage(resource.RUSAGE_SELF).ru_maxrss/1024.0/1024.0
    
    print ("%5.1f secs %5.1f MByte" % (time_elapsed,memMb))


src/convert_mtx.py
#!/usr/bin/env python
import os
import sys
import numpy
import pandas
import getopt

from scipy import io


def load_mtx(barcodes, genes, matrix, cell_map, min_count):
    title = os.path.basename(matrix)
    
    # quality check of input data first
    cell_map = pandas.read_csv(cell_map, sep='\t', index_col=0)
    
    for v in ['Cell Type', 'Sample']:
        if v not in cell_map.columns:
            sys.stderr.write('Error: %s cannot find %s in cell_map.\n' % (title, v))
            sys.exit(1)
    
    matrix = io.mmread(matrix)
    matrix = pandas.DataFrame.sparse.from_spmatrix(matrix)
    
    # assume first column of barcodes
    barcodes = pandas.read_csv(barcodes, sep='\t', header=None).iloc[:,0]
    
    # assume last column of genes is symbols
    genes = pandas.read_csv(genes, sep='\t', header=None).iloc[:, -1]
    
    assert matrix.shape[0] == genes.shape[0]
    assert matrix.shape[1] == barcodes.shape[0]
    
    matrix.index = genes
    matrix.columns = barcodes
    assert matrix.columns.value_counts().max() == 1
    
    # only focus on annotated cell barcodes
    common = matrix.columns.intersection(cell_map.index)
    if common.shape[0] == 0:
        sys.stderr.write('Error: %s no barcode annotated.\n' % title)
        sys.exit(1)
    
    matrix = matrix.loc[:, common]
    
    # jump bad cells, if any
    barcode_cnt = matrix.sum()
    if barcode_cnt.min() < min_count:
        matrix = matrix.loc[:, matrix.sum() >= min_count]

    # jump ambiguous genes, if any
    cnt_map = matrix.index.value_counts()
    if cnt_map.max() > 1:
        matrix = matrix.loc[cnt_map.loc[matrix.index] == 1]
    
    assert matrix.index.value_counts().max() == 1
    
    # remove empty genes
    matrix = matrix.loc[(matrix == 0).mean(axis=1) < 1]
    
    # make the column to CytoSig format
    matrix.columns = cell_map.loc[matrix.columns, 'Cell Type'] + '.' + cell_map.loc[matrix.columns, 'Sample'] + '.' + matrix.columns
    return matrix



def main():
    min_count = 1000
    min_ratio = 0.95
    
    input_lst = included = output = None
    
    prompt_msg = 'Usage:\nconvert_mtx.py -i <input list, separated by ,> -o <output prefix> -c <barcode count threshold, default: %d> -f <gene non-zero fraction threshold, default: %f> -s <cell subsets, default: leave empty>\n' % (min_count, min_ratio)
    
    try:
        opts, _ = getopt.getopt(sys.argv[1:], "hi:o:c:f:s:", [])
    
    except getopt.GetoptError:
        sys.stderr.write('Error input\n' + prompt_msg)
        sys.exit(2)
    
    if len(opts) == 0:
        sys.stderr.write('Please input some parameters or try: convert_mtx.py -h\n')
        sys.exit(2)
    
    for opt, arg in opts:
        if opt == '-h':
            print(prompt_msg)
            sys.exit()
        
        elif opt in ("-i"):
            input_lst = arg.split(',')
        
        elif opt in ("-s"):
            included = arg.split(',')
        
        elif opt in ("-o"):
            output = arg
        
        elif opt in ("-c"):
            try:
                min_count = int(arg)
            except:
                sys.stderr.write('count threshold %s is not a valid integer number.\n' % arg)
                sys.exit(1)
            
            if min_count < 100:
                sys.stderr.write('count threshold %d < 100. Please input a number >= 100\n' % min_count)
                sys.exit(2)
        
        elif opt in ("-f"):
            try:
                min_ratio = float(arg)
            except:
                sys.stderr.write('gene non-zero frequency threshold %s is not a valid float number.\n' % arg)
                sys.exit(1)
            
            if min_ratio > 1 or min_ratio <= 0:
                sys.stderr.write('count threshold %d < 100. Please input a number in (0, 1]\n' % min_count)
                sys.exit(2)
    
    
    if input_lst is None:
        sys.stderr.write('Please provide a input file list\n')
        sys.exit(2)
    
    if output is None:
        sys.stderr.write('Please provide an output prefix\n')
        sys.exit(2)        
    
    
    # only look at commonly profiled frequent genes across all subsets
    merge = []
    
    for fprefix in input_lst:
        barcodes = fprefix + 'barcodes.tsv.gz'
        genes = fprefix + 'genes.tsv.gz'
        matrix = fprefix + 'matrix.mtx.gz'
        cell_map = fprefix + 'cell_map.gz'
        
        for f in [barcodes, genes, matrix, cell_map]:
            if not os.path.exists(f):
                sys.stderr.write('Error: cannot find %s\n' % f)
                break
        
        matrix = load_mtx(barcodes, genes, matrix, cell_map, min_count)
        
        print(os.path.basename(fprefix), matrix.shape)
        
        merge.append(matrix)

    data = pandas.concat(merge, axis=1, join='inner')
    data = data.loc[(data == 0).mean(axis=1) < min_ratio]
    data = data.loc[:, data.sum() >= min_count]
    
    size_factor = 1E5/data.sum()
    data *= size_factor
    data = numpy.log2(data + 1)
    
    # always centralize on all cells, instead of included cells
    background = data.mean(axis=1)
    data = data.subtract(background, axis=0)

    if included is not None:
        data = data.loc[:, [v.split('.')[0] in included for v in data.columns]]
        
        if data.shape[0] == 0:
            sys.stderr.write('No cells included in %s.\n' % ','.join(included))
            sys.exit(2)
    
    print('data shape', data.shape)
    
    data.sparse.to_dense().to_pickle(output + '.pickle.gz', compression='gzip')
    
    return 0


if __name__ == '__main__':
    main()
